File: 48.rotate.c
#include "shdstd.h"

static inline void swap_i32(int *l, int *r)
{
	*l ^= *r;
	*r ^= *l;
	*l ^= *r;
}

static inline void swap_ptr(void **l, void **r)
{
	void *tmp = *l;
	*l = *r;
	*r = tmp;
}

void rotate(int **matrix, int size, int *col_size)
{
	int rows = size;
	int cols = col_size[0];

	for (int i = 0; i * 2 < size; ++i) {
		swap_ptr((void *)&matrix[i], (void *)&matrix[size-i-1]);
	}

	for (int r = 0; r < rows; ++r) {
		for (int c = r + 1; c < cols; ++c) {
			swap_i32(&matrix[r][c], &matrix[c][r]);
		}
	}
}

File: 48.rotate.go
package main

func rotate(matrix [][]int) {
	rows := len(matrix)
	cols := len(matrix[0])

	for left := 0; left*2 < rows; left++ {
		right := rows - left - 1
		matrix[left], matrix[right] = matrix[right], matrix[left]
	}

	for r := 0; r < rows; r++ {
		for c := r + 1; c < cols; c++ {
			matrix[r][c], matrix[c][r] = matrix[c][r], matrix[r][c]
		}
	}
}
